### PR TITLE
Remove .T as shortcut for transpose()

### LINKED ISSUE #1232: Remove the `Dataset.T` property?
The fact that `ds.T` transposes a dataset rather than returning `ds['T']` is a repeated point of confusion:
http://stackoverflow.com/questions/41879269/xarray-dataarray-opens-as-a-dataset/41885639#41885639
http://stackoverflow.com/questions/40270929/how-to-read-a-variable-called-t-with-xarray

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #1232
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API (remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later)

### CODE REVIEW

**Comment by jhamman** [comment]
For posterity, what did we decide to do with the `DataArray.T` property?

**Comment by dcherian** [comment]
> For posterity, what did we decide to do with the DataArray.T property?

I'm not sure but I've fixed the whats-new entry to say that only Dataset.T has been removed currently.

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -917,10 +917,7 @@ def _item_sources(self):
                 LevelCoordinatesSource(self)]
 
     def __dir__(self):
-        # In order to suppress a deprecation warning in Ipython autocompletion
-        # .T is explicitly removed from __dir__. GH: issue 1675
         d = super(Dataset, self).__dir__()
-        d.remove('T')
         return d
```
Comment (shoyer) [inline xarray/core/dataset.py]: You can drop this method entirely in favor of the super class method.

**Comment by shoyer** [comment]
> For posterity, what did we decide to do with the `DataArray.T` property?

I think this is arguably still useful/expected since DataArray is modeled on a NumPy array. But I could probably be convinced otherwise...

**Comment by dcherian** [comment]
> I think this is arguably still useful/expected since DataArray is modeled on a NumPy array.

I agree

**Comment by jhamman** [comment]
Feeling like playing devils advocate here. I would expect we'll run into similar confusion with the `DataArray` transpose property. 


```Python

>>> da = xr.DataArray([1, 2, 3, 4], dims=('time', ), coords={'time': pd.date_range('2018-10-24', periods=4)})
>>> da.time

<xarray.DataArray 'time' (time: 4)>
array(['2018-10-24T00:00:00.000000000', '2018-10-25T00:00:00.000000000',
       '2018-10-26T00:00:00.000000000', '2018-10-27T00:00:00.000000000'],
      dtype='datetime64[ns]')
Coordinates:
  * time     (time) datetime64[ns] 2018-10-24 2018-10-25 2018-10-26 2018-10-27

>>> da = xr.DataArray([1, 2, 3, 4], dims=('T', ), coords={'T': pd.date_range('2018-10-24', periods=4)})
>>> da.T

<xarray.DataArray (T: 4)>
array([1, 2, 3, 4])
Coordinates:
  * T        (T) datetime64[ns] 2018-10-24 2018-10-25 2018-10-26 2018-10-27
```

**Comment by max-sixty** [comment]
I'm +0.2 for merging this, but _at least_ issuing a `DeprecationWarning` before removing from `DataArray.T` (with no strong view on whether we do remove it)

**Comment by jhamman** [comment]
Sure, that's a good point. I just wanted to illustrate that these are the same problems and that only removing the property from the Dataset may lead to continued confusion. @shoyer thoughts?

**Comment by shoyer** [comment]
I suppose the case for removing `DataArray.T` is that you don't really need a shortcut for transposing arrays in xarray, since most functions don't rely on dimension order. And of course the real motivation: some weather/climate scientists like to use a variable named `T` for time :).

The counter-argument would be that `T` is found on most other numpy array-like types, including `pandas.DataFrame`. I've definitely used `.T` as a convenient shortcut myself in some cases.

I guess that puts me at -0 for deprecating `DataArray.T`. But we can merge this already, and leave that for a separate discussion.

**Comment by max-sixty** [comment]
>  I just wanted to illustrate that these are the same problems and that only removing the property from the Dataset may lead to continued confusion. 

I think that's a good point. There are a few methods on `DataArray`s but not `Dataset`s, but not many:

```python
In [4]: {x for x in set(dir(xr.DataArray)) - set(dir(xr.Dataset)) if not x.startswith("_")}
Out[4]:
{'data',
 'dot',
 'dt',
 'dtype',
 'from_cdms2',
 'from_iris',
 'from_series',
 'get_axis_num',
 'item',
 'name',
 'ndim',
 'plot',
 'searchsorted',
 'shape',
 'size',
 'to_cdms2',
 'to_dataset',
 'to_index',
 'to_iris',
 'to_masked_array',
 'to_pandas',
 'to_series',
 'variable'}
```

Regardless, sounds like @shoyer thinks we should merge (and I'm probably still +0.1)

File: doc/whats-new.rst [hunk 2]
```diff
@@ -33,6 +33,8 @@ v0.11.0 (unreleased)
 Breaking changes
 ~~~~~~~~~~~~~~~~
 
+- ``Dataset.T`` has been removed as a shortcut for :py:meth:`transpose()`.
```
Comment (max-sixty) [inline doc/whats-new.rst]: Generally we don't put `()` in the docs - is this our convention or will the docs not link?

Comment (dcherian) [reply doc/whats-new.rst]: good catch

Comment (max-sixty) [reply doc/whats-new.rst]: It's also in the line below! 
I really don't think it matters if it's only cosmetic, though

Comment (dcherian) [reply doc/whats-new.rst]: image

**Pull request merged**